internal/domain/user.go

```diff
@@ -0,0 +1,136 @@
+package domain
+
+import (
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/go-playground/validator/v10"
+	"github.com/google/uuid"
+)
+
+type UserAggregate struct {
+	id                    uuid.UUID `validate:"required"`
+	Username              string    `validate:"required,min=2,max=50"`
+	Email                 string    `validate:"required,email"`
+	isVerified            bool
+	verifiedAt            *time.Time
+	lastPasswordChangedAt *time.Time
+	registeredAt          time.Time `validate:"required"`
+	hashedPassword        []byte    `validate:"required"`
+	isSuperUser           bool
+}
+
+func (u *UserAggregate) GetIdentifier() uuid.UUID {
+	return u.id
+}
+
+func (u *UserAggregate) Validate() error {
+	valdidate := validator.New()
+	return valdidate.Struct(u)
+}
+
+func (u *UserAggregate) ChangedEmail(newEmail string) error {
+	validate := validator.New()
+
+	if err := validate.Var(newEmail, "required,email"); err != nil {
+		return err
+	}
+
+	u.Email = newEmail
+	return nil
+}
+
+func (u *UserAggregate) VerifyPassword(password string) bool {
+	svc := NewPasswordDomainService()
+	return svc.Verify(u.hashedPassword, password)
+}
+
+func (u *UserAggregate) setNewPassword(password string) error {
+	svc := NewPasswordDomainService()
+
+	hashedPassword, err := svc.Hash(password)
+
+	if err != nil {
+		return fmt.Errorf("error during hashed password: %v", err)
+	}
+
+	nowUTC := time.Now().UTC()
+	u.hashedPassword = hashedPassword
+	u.lastPasswordChangedAt = &nowUTC
+	return nil
+}
+
+func (u *UserAggregate) ChangePassword(password string) error {
+	return u.setNewPassword(password)
+}
+
+func (u *UserAggregate) ResetPassword() (string, error) {
```

Comment:
It's interesting to see "reset password" that you'd expect user to enter a new password but here you just overwrite their old with some random one 😅 I am assuming that I don't know the use case.

Reply:
It will be a functionality that can be performed by a super user in case of account theft along with the email. A normal user won't be able to use this functionality :)